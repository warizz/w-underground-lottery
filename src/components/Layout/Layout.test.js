import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { Layout as Uut } from './';

describe('snapshot testing', () => {
  it('should match snapshot when no props', () => {
    const wrapper = shallow(<Uut />);
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

  it('should display placeholder when fetching initial data', () => {
    const wrapper = shallow(<Uut />);
    wrapper.setState({ isFetchingInitialData: true });
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });

  it('should render content when have children', () => {
    const wrapper = shallow(
      <Uut>
        <div>test</div>
      </Uut>
    );
    const tree = toJson(wrapper);
    expect(tree).toMatchSnapshot();
  });
});

describe('functional testing', () => {
  test('closeSnackbar', () => {
    const props = {
      setAlert: jest.fn(),
    };
    const wrapper = shallow(<Uut {...props} />);
    wrapper.instance().closeSnackbar();
    expect(props.setAlert).toHaveBeenCalledWith('');
  });

  describe('fetchInitialData', () => {
    test('on success', async () => {
      const mockUser = {};
      const mockCurrentPeriod = {};
      const props = {
        proxy: {
          getCurrentPeriod: async () => mockCurrentPeriod,
          getUser: async () => mockUser,
        },
        setCurrentPeriod: jest.fn(),
        setUser: jest.fn(),
      };
      const wrapper = shallow(<Uut {...props} />);
      await wrapper.instance().fetchInitialData();
      expect(props.setCurrentPeriod).toHaveBeenCalledWith(mockCurrentPeriod);
      expect(props.setUser).toHaveBeenCalledWith(mockUser);
    });

    test('on error', async () => {
      const props = {
        proxy: {
          getCurrentPeriod: async () => Promise.rejct(),
          getUser: async () => Promise.rejct(),
        },
        setCurrentPeriod: jest.fn(),
        setUser: jest.fn(),
      };
      const wrapper = shallow(<Uut {...props} />);
      await wrapper.instance().fetchInitialData();
      expect(props.setCurrentPeriod).toHaveBeenCalledTimes(0);
      expect(props.setUser).toHaveBeenCalledTimes(0);
    });

    test('on 401 error', async () => {
      const mockError = new Error();
      mockError.response = {
        status: 401,
      };
      const props = {
        cookieManager: {
          removeToken: jest.fn(),
        },
        proxy: {
          getCurrentPeriod: async () => Promise.reject(mockError),
          getUser: async () => Promise.reject(mockError),
        },
        router: {
          push: jest.fn(),
        },
        setCurrentPeriod: jest.fn(),
        setUser: jest.fn(),
      };
      const wrapper = shallow(<Uut {...props} />);
      await wrapper.instance().fetchInitialData();
      expect(props.setCurrentPeriod).toHaveBeenCalledTimes(0);
      expect(props.setUser).toHaveBeenCalledTimes(0);
      expect(props.cookieManager.removeToken).toHaveBeenCalled();
      expect(props.router.push).toHaveBeenCalledWith('/log-in');
    });
  });
});
