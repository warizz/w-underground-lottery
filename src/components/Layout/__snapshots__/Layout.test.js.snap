// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot testing should display placeholder when fetching initial data 1`] = `
<styled.div>
  <h1>
    Initiating your data ...
  </h1>
</styled.div>
`;

exports[`snapshot testing should match snapshot when no props 1`] = `
<div>
  <Toolbar
    onClickMainButton={[Function]}
    pageName=""
  />
</div>
`;

exports[`snapshot testing should render content when have children 1`] = `
<div>
  <Toolbar
    onClickMainButton={[Function]}
    pageName=""
  />
  <styled.div>
    <Snackbar
      onClose={[Function]}
      text=""
      timer={1000}
    />
    <div>
      test
    </div>
  </styled.div>
</div>
`;
